src/code/z_play.c

```diff
@@ -1,81 +1,1545 @@
+HiresoStruct sHireso;
+u8 sBombersNotebookOpen;
+u8 sMotionBlurStatus;
+
+typedef enum {
+    /* 0 */ MOTION_BLUR_STOPPED,
+    /* 1 */ MOTION_BLUR_INITALIZING,
+    /* 2 */ MOTION_BLUR_RUNNING
+} MotionBlurStatus;
+
+void Play_DrawMotionBlur(PlayState* this) {
+    GraphicsContext* gfxCtx = this->state.gfxCtx;
+    s32 alpha;
+    Gfx* gfx;
+    Gfx* dlistHead;
+
+    if (R_MOTION_BLUR_PRIORITY_ENABLED) {
+        alpha = R_MOTION_BLUR_PRIORITY_ALPHA;
+
+        if (sMotionBlurStatus == MOTION_BLUR_STOPPED) {
+            sMotionBlurStatus = MOTION_BLUR_INITALIZING;
+        }
+    } else if (R_MOTION_BLUR_ENABLED) {
+        alpha = R_MOTION_BLUR_ALPHA;
+
+        if (sMotionBlurStatus == MOTION_BLUR_STOPPED) {
+            sMotionBlurStatus = MOTION_BLUR_INITALIZING;
+        }
+    } else {
+        alpha = 0;
+        sMotionBlurStatus = MOTION_BLUR_STOPPED;
+    }
+
+    if (sMotionBlurStatus != MOTION_BLUR_STOPPED) {
+        OPEN_DISPS(gfxCtx);
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_play/func_80165460.s")
+        dlistHead = POLY_OPA_DISP;
+        gfx = Graph_GfxPlusOne(dlistHead);
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_play/func_80165608.s")
+        gSPDisplayList(OVERLAY_DISP++, gfx);
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_play/func_80165630.s")
+        this->pauseBgPreRender.fbuf = gfxCtx->curFrameBuffer;
+        this->pauseBgPreRender.fbufSave = this->unk_18E64;
+
+        if (sMotionBlurStatus == MOTION_BLUR_RUNNING) {
+            func_80170AE0(&this->pauseBgPreRender, &gfx, alpha);
+        } else {
+            sMotionBlurStatus = MOTION_BLUR_RUNNING;
+        }
+
+        func_801705B4(&this->pauseBgPreRender, &gfx);
+
+        gSPEndDisplayList(gfx++);
+
+        Graph_BranchDlist(dlistHead, gfx);
+
+        POLY_OPA_DISP = gfx;
+
+        CLOSE_DISPS(gfxCtx);
+    }
+}
+
+void Play_InitMotionBlur(void) {
+    R_MOTION_BLUR_ENABLED = false;
+    R_MOTION_BLUR_PRIORITY_ENABLED = false;
+    sMotionBlurStatus = MOTION_BLUR_STOPPED;
+}
+
+void Play_DestroyMotionBlur(void) {
+    R_MOTION_BLUR_ENABLED = false;
+    R_MOTION_BLUR_PRIORITY_ENABLED = false;
+    sMotionBlurStatus = MOTION_BLUR_STOPPED;
+}
+
+void Play_SetMotionBlurAlpha(u32 alpha) {
+    R_MOTION_BLUR_ALPHA = alpha;
+}
+
+void Play_EnableMotionBlur(u32 alpha) {
+    R_MOTION_BLUR_ALPHA = alpha;
+    R_MOTION_BLUR_ENABLED = true;
+}
+
+void Play_DisableMotionBlur(void) {
+    R_MOTION_BLUR_ENABLED = false;
+}
+
+void Play_ConvertRgba16ToIntensityImage(void* destI, u16* srcRgba16, s32 rgba16Width, s32 pixelLeft, s32 pixelTop,
+                                        s32 pixelRight, s32 pixelBottom, s32 bitDepth) {
+    s32 i;
+    s32 j;
+    u32 pixel;
+    u32 r;
+    u32 g;
+    u32 b;
+
+    switch (bitDepth) {
+        case 4: {
+            u8* destI4 = (u8*)destI;
+            u32 upper;
+            u32 lower;
+
+            for (i = pixelTop; i <= pixelBottom; i++) {
+                for (j = pixelLeft; j <= pixelRight; j += 2) {
```

Comment:
@Thar0 reminds me that it's `G_IM_SIZ_16b_BYTES`.

Reply:
Yes, it's processing 2 at a time because it's processing to make one byte of the output at a time, right? If you were squashing an RGBA32 in you'd want to do it 4 at a time, etc.

Reply:
I'm sure that is quite right? Both I5 and I8 are 1 byte or less, but they still only process 1 pixel at a time.

Reply:
I meant in that the loops in this function operate one byte of the image at a time, so you get
- 4: 2 -> 1
- 5,8: 1 -> 1
- 16: 2 -> 2 (treated as 1)

and these factors can be expressed as ratios, although I suppose the issue here is that the ratios are the wrong way up and so don't really work as integers: if `G_IM_SIZ_4b_BYTES` was 1/2 it would work.

Reply:
Okay I can see that. But as you mention `G_IM_SIZ_4b_BYTES` would be the correct define, but only if it was `1/2`. For that reason, and the fact it would be weird to do it for the other ones as well, I think it is just best to leave it as 2.